test/data_map_test.exs

```diff
@@ -0,0 +1,162 @@
+defmodule DataMapTest do
+  use PowerAssert
+
+  test "Insert empty row must fail" do
+    c1 = %Masdb.Schema.Column{is_pk: true,  name: "c1", type: :int}
+    c2 = %Masdb.Schema.Column{is_pk: false, name: "c2", type: :int}
+    c3 = %Masdb.Schema.Column{is_pk: false, name: "c3", type: :int}
+
+    timestamp = Masdb.Timestamp.get_timestamp()
+    inserted_value = %{}
+    inserted_schema = %Masdb.Schema{name: "foo", replication_factor: 0, columns: [c1, c2, c3]}
+    inserted_nodeid = "foo@127.0.0.1"
+
+    map = %Masdb.Data.Map{
+      node_id: inserted_nodeid,
+      last_update_time: timestamp,
+      last_sync_time: timestamp,
+      map: %{"foo" => %Masdb.Data.Table{}}
+    }
+
+    {_, {:cannot_insert_empty_row, _}} = Masdb.Data.Map.insert(map, inserted_schema, inserted_value)
+  end
+
+  test "Insert invalid row must fail" do
+    c1 = %Masdb.Schema.Column{is_pk: true,  name: "c1", type: :int}
+    c2 = %Masdb.Schema.Column{is_pk: false, name: "c2", type: :int}
+    c3 = %Masdb.Schema.Column{is_pk: false, name: "c3", type: :int}
+
+    timestamp = Masdb.Timestamp.get_timestamp()
+    inserted_value = %{"c1" => "val1", "c2" => "val2", "c3" => "val3", "invalid_col" => "val"}
+    inserted_schema = %Masdb.Schema{name: "foo", replication_factor: 0, columns: [c1, c2, c3]}
+    inserted_nodeid = "foo@127.0.0.1"
+
+    map = %Masdb.Data.Map{
+      node_id: inserted_nodeid,
+      last_update_time: timestamp,
+      last_sync_time: timestamp,
+      map: %{"foo" => %Masdb.Data.Table{}}
+    }
+
+    {_, {:col_doesnt_exists, _}} = Masdb.Data.Map.insert(map, inserted_schema, inserted_value)
+  end
+
+  test "Insert duplicate pk in datamap must fail" do
+    c1 = %Masdb.Schema.Column{is_pk: true,  name: "c1", type: :int}
+    c2 = %Masdb.Schema.Column{is_pk: false, name: "c2", type: :int}
+    c3 = %Masdb.Schema.Column{is_pk: false, name: "c3", type: :int}
+
+    timestamp = Masdb.Timestamp.get_timestamp()
+    inserted_value = %{"c1" => "val1", "c2" => "val2", "c3" => "val3"}
+    inserted_schema = %Masdb.Schema{name: "foo", replication_factor: 0, columns: [c1, c2, c3]}
+    inserted_nodeid = "foo@127.0.0.1"
+
+    duplicated_row_id = inserted_nodeid <> to_string(timestamp.unique_integer)
+
+    map = %Masdb.Data.Map{
+      node_id: inserted_nodeid,
+      last_update_time: timestamp,
+      last_sync_time: timestamp,
+      map: %{"foo" => %Masdb.Data.Table {
```

Comment:
It could be more understandable, but I think this way we ensure that the function work with a given state in input. It's a bit more white box this way